INT-4060: FTP Gateway: Add `NLST` and `workDir`

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4060

* Add `NLST` command to the `AbstractRemoteFileOutboundGateway` to perform
`listNames` on the target session.
Useful in case of server doesn't allow to perform `LS` or the names set is
sufficient for application requirements
* Add `workingDirExpression` to the `FtpOutboundGateway` to allow to perform
`FtpClient.changeWorkingDirectory()` based on the current request message
* Change `slf4j-log4j12` to the `testCompile` -
the FTP tests fail in the IDE with `ClassNotFoundException`

artembilan committed:
INT-4060: FTP Gateway: Add `NLST` and `workDir`

JIRA: https://jira.spring.io/browse/INT-4060

* Add `NLST` command to the `AbstractRemoteFileOutboundGateway` to perform
`listNames` on the target session.
Useful in case of server doesn't allow to perform `LS` or the names set is
sufficient for application requirements
* Add `workingDirExpression` to the `FtpOutboundGateway` to allow to perform
`FtpClient.changeWorkingDirectory()` based on the current request message
* Change `slf4j-log4j12` to the `testCompile` -
the FTP tests fail in the IDE with `ClassNotFoundException`

artembilan commented:
If you are fine with the solution, TODO:
* add `nlst` to the XSD
* wrap more commands to the `doInWorkingDirectory()` in the `FtpOutboundGateway`
* Documentation

garyrussell commented on spring-integration-ftp/src/test/java/org/springframework/integration/ftp/outbound/FtpServerOutboundTests-context.xml:
```diff
@@ -178,8 +178,8 @@
-							  command-options="-1"
+							  command="nlst"
+							  working-dir-expression="'ftpSource'"
 							  reply-channel="output"/>
```
We should really test both forms.

artembilan commented on spring-integration-ftp/src/main/java/org/springframework/integration/ftp/gateway/FtpOutboundGateway.java:
```diff
@@ -151,9 +185,75 @@ protected long getModified(FTPFile file) {
+			if (workingDirExpression != null) {
+				currentWorkingDirectory = ftpClient.printWorkingDirectory();
+				String workingDir = workingDirExpression.getValue(this.evaluationContext, message, String.class);
+				ftpClient.changeWorkingDirectory(workingDir);
```
`null` is fine for FTP: http://docs.spring.io/spring-integration/reference/html/ftp.html#ftp-inbound :
> Starting with version 4.3, the `remote-directory/remote-directory-expression` attributes can be omitted assuming `null`. In this case, according to the FTP protocol, the Client working directory is used as a default remote directory.

garyrussell replied:
True, but if `null`, we can avoid 2 round-trips to the server (change and change back noops).

artembilan replied:
Well, we can't assume that user uses some other working dir by default and decided for some messages to switch to his/her root. And, of course, then come back after the operation. Remember `AbstractFtpSessionFactory.postProcessClientBeforeConnect/postProcessClientAfterConnect` ?

> The working directory can be set via the FTPClient.changeWorkingDirectory() function when you extend the DefaultFtpSessionFactory and implement postProcessClientAfterConnect() callback.

artembilan committed:
Address PR comments

* Add `nlst` to XSD config
* Reinstate `ls -1` test-case for the `FtpServerOutboundTests`
* wrap more commands to the `doInWorkingDirectory()`

artembilan commented:
Still need documentation.

There is a question, though, regarding `PUT` and `MPUT`.
They are based on the hard work in the `RemoteFileTemplate.send()`.
What is bad there that each iteration in the `doMput()` may perform that `send` on a new `Session` (and new Client respectively).
How about to add `ThreadLocal<Session<F>>`  trick to the `RemoteFileTemplate` let it to reuse an instance for the current invocation? Plus with that we will be able to insert `workingDirectory`.
WDYT?

artembilan committed:
Fix `MV` command in the `AbstractRemoteFileOutboundGateway`

garyrussell commented:
I assume you mean similar to the `RabbitOperations.invoke()` - if so, yes, I agree that would be a good pattern for those methods.

artembilan commented:
M-m-m. No, `MGET` is fine. It is already based on the dedicated `session`:
```
List<File> payload = this.remoteFileTemplate.execute(session ->
				mGet(requestMessage, session, remoteDir, remoteFilename));
```

artembilan committed:
* Implement `RemoteFileOperations#invoke(OperationsCallback<F, T>)`
for thread-bound `session`s
* Use a new `invoke()` for `put()` and `mPut()` commands in the `AbstractRemoteFileOutboundGateway`
* Add delegation for the `put()` and `mPut()` commands in the `FtpOutboundGateway`
* Add DSL support for the `workingDirExpression` and add `whats-new.adoc` note

artembilan commented:
Still need general doc for the new API, but there is already a plenty of changes to review.
Will document tomorrow.

artembilan committed:
Document changes

garyrussell commented on spring-integration-file/src/main/java/org/springframework/integration/file/remote/RemoteFileTemplate.java:
```diff
@@ -398,15 +407,25 @@ public boolean get(final String remotePath, final InputStreamCallback callback)
 	public Session<F> getSession() {
-		return this.sessionFactory.getSession();
+		if (this.activeTemplateCallbacks.get() > 0) {
+			return this.contextSessions.get();
```
Need to check for null.

garyrussell commented on spring-integration-file/src/main/java/org/springframework/integration/file/remote/RemoteFileTemplate.java:
```diff
@@ -420,7 +439,7 @@ public boolean get(final String remotePath, final InputStreamCallback callback)
 		}
 		finally {
-			if (session != null) {
+			if (this.activeTemplateCallbacks.get() == 0 && session != null) {
```
This isn't right - it might be another thread that's running in an invoke.

See `invokeScope` variable in rabbit template's doExecute

garyrussell commented on spring-integration-file/src/main/java/org/springframework/integration/file/remote/RemoteFileTemplate.java:
```diff
@@ -434,6 +453,28 @@ public boolean get(final String remotePath, final InputStreamCallback callback)
+		if (contextSession == null) {
+			this.contextSessions.set(this.sessionFactory.getSession());
+		}
+		this.activeTemplateCallbacks.incrementAndGet();
```
We should disallow nested invokes - see `RabbitTemplate`.

artembilan replied:
Why? I don't see any problems here with reusing already stored in the `ThreadLocal` session. 
I might agree about your comment in the `finally` block: have had similar doubts. But then decided to leave it somehow... Let me recall what was my conclusion.

We might consider to use additional `ThreadLocal` counter to track the recursion though...

garyrussell replied:
I just think it adds unnecessary complications (recursive calls to invoke).

garyrussell replied:
The `activeTemplateCallbacks` is simply an optimization to avoid accessing the thread local if no invokes are in process.

artembilan replied:
I leave the logic in this method as is. That isn't so hard to track recursion here. Plus check our `MPUT` impl, especially when it is recursive as well. In the end it call `doPut()` for each file, which is based on the `invoke()` as well, since we have there not only `send()` operation, but `doChmod()`, too.

garyrussell reviewed (changes requested):
That's all for this round.

artembilan committed:
Address some PR comments:

* Add `invokeScope` variable to `execute` to track `ThreadLocal` session or not
* Check for `null` in the `getSession()` and fallback to regular
`sessionFactory.getSession()`.
Most likely the `invoke()` is called from other thread

garyrussell commented:
Doc polishing for your approval https://github.com/garyrussell/spring-integration/commit/cd1f6809f59dbf46b46e7897ccbfaaded476cf02

garyrussell commented:
Merged as 3daeaab1b4b20d868a8b07d6b0813023a492e56c
